File: dbflow-rx/src/main/java/com/raizlabs/android/dbflow/rx/structure/RXModelAdapter.java
package com.raizlabs.android.dbflow.rx.structure;

import com.raizlabs.android.dbflow.config.FlowManager;
import com.raizlabs.android.dbflow.structure.ModelAdapter;
import com.raizlabs.android.dbflow.structure.database.DatabaseWrapper;

import java.util.Collection;
import java.util.concurrent.Callable;

import rx.Completable;
import rx.Single;

/**
 * Description: Wraps most {@link ModelAdapter} modification operations into RX-style constructs.
 */
public class RXModelAdapter<T> extends RXRetrievalAdapter<T> {

    public static <T> RXModelAdapter<T> from(ModelAdapter<T> modelAdapter) {
        return new RXModelAdapter<>(modelAdapter);
    }

    public static <T> RXModelAdapter<T> from(Class<T> table) {
        return new RXModelAdapter<>(table);
    }

    private final ModelAdapter<T> modelAdapter;

    RXModelAdapter(ModelAdapter<T> modelAdapter) {
        super(modelAdapter);
        this.modelAdapter = modelAdapter;
    }

    public RXModelAdapter(Class<T> table) {
        this(FlowManager.getModelAdapter(table));
    }

    public Single<Boolean> save(final T model) {
        return Single.fromCallable(new Callable<Boolean>() {
            @Override
            public Boolean call() throws Exception {
                return modelAdapter.save(model);
            }
        });
    }

    public Single<Boolean> save(final T model, final DatabaseWrapper databaseWrapper) {
        return Single.fromCallable(new Callable<Boolean>() {
            @Override
            public Boolean call() throws Exception {
                return modelAdapter.save(model, databaseWrapper);
            }
        });
    }

    public Completable saveAll(final Collection<T> models) {
        return Completable.fromCallable(new Callable<Void>() {
            @Override
            public Void call() throws Exception {
                modelAdapter.saveAll(models);
                return null;
            }
        });
    }

    public Completable saveAll(final Collection<T> models, final DatabaseWrapper databaseWrapper) {
        return Completable.fromCallable(new Callable<Void>() {
            @Override
            public Void call() throws Exception {
                modelAdapter.saveAll(models, databaseWrapper);
                return null;
            }
        });
    }

    public Single<Long> insert(final T model) {
        return Single.fromCallable(new Callable<Long>() {
            @Override
            public Long call() throws Exception {
                return modelAdapter.insert(model);
            }
        });
    }

    public Single<Long> insert(final T model, final DatabaseWrapper databaseWrapper) {
        return Single.fromCallable(new Callable<Long>() {
            @Override
            public Long call() throws Exception {
                return modelAdapter.insert(model, databaseWrapper);
            }
        });
    }

    public Completable insertAll(final Collection<T> models) {
        return Completable.fromCallable(new Callable<Void>() {
            @Override
            public Void call() throws Exception {
                modelAdapter.insertAll(models);
                return null;
            }
        });
    }

    public Completable insertAll(final Collection<T> models,
                                  final DatabaseWrapper databaseWrapper) {
        return Completable.fromCallable(new Callable<Void>() {
            @Override
            public Void call() throws Exception {
                modelAdapter.insertAll(models, databaseWrapper);
                return null;
            }
        });
    }

    public Single<Boolean> update(final T model) {
        return Single.fromCallable(new Callable<Boolean>() {
            @Override
            public Boolean call() throws Exception {
                return modelAdapter.update(model);
            }
        });
    }

    public Completable update(final T model, final DatabaseWrapper databaseWrapper) {
        return Completable.fromCallable(new Callable<Boolean>() {
            @Override
            public Boolean call() throws Exception {
                return modelAdapter.update(model, databaseWrapper);
            }
        });
    }

    public Completable updateAll(final Collection<T> models) {
        return Completable.fromCallable(new Callable<Void>() {
            @Override
            public Void call() throws Exception {
                modelAdapter.updateAll(models);
                return null;
            }
        });
    }

    public Completable updateAll(final Collection<T> models, final DatabaseWrapper databaseWrapper) {
        return Completable.fromCallable(new Callable<Void>() {
            @Override
            public Void call() throws Exception {
                modelAdapter.updateAll(models, databaseWrapper);
                return null;
            }
        });
    }

    public Single<Boolean> delete(final T model) {
        return Single.fromCallable(new Callable<Boolean>() {
            @Override
            public Boolean call() throws Exception {
                return modelAdapter.delete(model);
            }
        });
    }

    public Single<Boolean> delete(final T model, final DatabaseWrapper databaseWrapper) {
        return Single.fromCallable(new Callable<Boolean>() {
            @Override
            public Boolean call() throws Exception {
                return modelAdapter.delete(model, databaseWrapper);
            }
        });
    }

    public Completable deleteAll(final Collection<T> models) {
        return Completable.fromCallable(new Callable<Void>() {
            @Override
            public Void call() throws Exception {
                modelAdapter.deleteAll(models);
                return null;
            }
        });
    }

    public Completable deleteAll(final Collection<T> models, final DatabaseWrapper databaseWrapper) {
        return Completable.fromCallable(new Callable<Void>() {
            @Override
            public Void call() throws Exception {
                modelAdapter.deleteAll(models, databaseWrapper);
                return null;
            }
        });
    }
}

File: dbflow-rx/src/main/java/com/raizlabs/android/dbflow/rx/structure/RXModelAdapter.kt
package com.raizlabs.android.dbflow.rx.structure

import com.raizlabs.android.dbflow.config.FlowManager
import com.raizlabs.android.dbflow.structure.ModelAdapter
import com.raizlabs.android.dbflow.structure.database.DatabaseWrapper
import rx.Completable
import rx.Single

/**
 * Description: Wraps most [ModelAdapter] modification operations into RX-style constructs.
 */
class RXModelAdapter<T : Any> internal constructor(private val modelAdapter: ModelAdapter<T>)
    : RXRetrievalAdapter<T>(modelAdapter) {

    constructor(table: Class<T>) : this(FlowManager.getModelAdapter<T>(table))

    fun save(model: T): Single<Boolean> = Single.fromCallable { modelAdapter.save(model) }

    fun save(model: T, databaseWrapper: DatabaseWrapper): Single<Boolean> =
            Single.fromCallable { modelAdapter.save(model, databaseWrapper) }

    fun saveAll(models: Collection<T>): Completable = Completable.fromCallable {
        modelAdapter.saveAll(models)
        null
    }

    fun saveAll(models: Collection<T>, databaseWrapper: DatabaseWrapper): Completable =
            Completable.fromCallable {
                modelAdapter.saveAll(models, databaseWrapper)
                null
            }

    fun insert(model: T): Single<Long> = Single.fromCallable { modelAdapter.insert(model) }

    fun insert(model: T, databaseWrapper: DatabaseWrapper): Single<Long> =
            Single.fromCallable { modelAdapter.insert(model, databaseWrapper) }

    fun insertAll(models: Collection<T>): Completable = Completable.fromCallable {
        modelAdapter.insertAll(models)
        null
    }

    fun insertAll(models: Collection<T>,
                  databaseWrapper: DatabaseWrapper): Completable = Completable.fromCallable {
        modelAdapter.insertAll(models, databaseWrapper)
        null
    }

    fun update(model: T): Single<Boolean> = Single.fromCallable { modelAdapter.update(model) }

    fun update(model: T, databaseWrapper: DatabaseWrapper): Single<Boolean> =
            Single.fromCallable { modelAdapter.update(model, databaseWrapper) }

    fun updateAll(models: Collection<T>): Completable = Completable.fromCallable {
        modelAdapter.updateAll(models)
        null
    }

    fun updateAll(models: Collection<T>, databaseWrapper: DatabaseWrapper): Completable =
            Completable.fromCallable {
                modelAdapter.updateAll(models, databaseWrapper)
                null
            }

    fun delete(model: T): Single<Boolean> = Single.fromCallable { modelAdapter.delete(model) }

    fun delete(model: T, databaseWrapper: DatabaseWrapper): Single<Boolean> =
            Single.fromCallable { modelAdapter.delete(model, databaseWrapper) }

    fun deleteAll(models: Collection<T>): Completable = Completable.fromCallable {
        modelAdapter.deleteAll(models)
        null
    }

    fun deleteAll(models: Collection<T>, databaseWrapper: DatabaseWrapper): Completable =
            Completable.fromCallable {
                modelAdapter.deleteAll(models, databaseWrapper)
                null
            }

    companion object {

        @JvmStatic
        fun <T : Any> from(modelAdapter: ModelAdapter<T>): RXModelAdapter<T> =
                RXModelAdapter(modelAdapter)

        @JvmStatic
        fun <T : Any> from(table: Class<T>): RXModelAdapter<T> = RXModelAdapter(table)
    }
}
